java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/MarkOccurrencesHighlighterBase.java

```diff
@@ -499,42 +501,61 @@ protected final synchronized void resume() {
+        StatementTree targetStatementTree = null;
+        ExpressionTree targetExpressionTree = null;
+        if (target instanceof StatementTree) {
+            targetStatementTree = target.getKind() == Kind.LABELED_STATEMENT ?  ((LabeledStatementTree) target).getStatement() : (StatementTree) target;
+        } else if (target instanceof ExpressionTree) {
+            targetExpressionTree = (ExpressionTree) target;
+        }
         Tree block = null;
-
-        switch (statement.getKind()) {
-            case SWITCH:
-                block = statement;
-                break;
-            case WHILE_LOOP:
-                if (((WhileLoopTree) statement).getStatement().getKind() == Kind.BLOCK)
-                    block = ((WhileLoopTree) statement).getStatement();
-                break;
-            case FOR_LOOP:
-                if (((ForLoopTree) statement).getStatement().getKind() == Kind.BLOCK)
-                    block = ((ForLoopTree) statement).getStatement();
-                break;
-            case ENHANCED_FOR_LOOP:
-                if (((EnhancedForLoopTree) statement).getStatement().getKind() == Kind.BLOCK)
-                    block = ((EnhancedForLoopTree) statement).getStatement();
-                break;
-            case DO_WHILE_LOOP:
-                if (((DoWhileLoopTree) statement).getStatement().getKind() == Kind.BLOCK)
-                    block = ((DoWhileLoopTree) statement).getStatement();
-                break;
+        
+        if (targetStatementTree != null) {
+            switch (targetStatementTree.getKind()) {
+                case SWITCH:
+                    block = targetStatementTree;
+                    break;
+                case WHILE_LOOP:
+                    if (((WhileLoopTree) targetStatementTree).getStatement().getKind() == Kind.BLOCK)
+                        block = ((WhileLoopTree) targetStatementTree).getStatement();
+                    break;
+                case FOR_LOOP:
+                    if (((ForLoopTree) targetStatementTree).getStatement().getKind() == Kind.BLOCK)
+                        block = ((ForLoopTree) targetStatementTree).getStatement();
+                    break;
+                case ENHANCED_FOR_LOOP:
+                    if (((EnhancedForLoopTree) targetStatementTree).getStatement().getKind() == Kind.BLOCK)
+                        block = ((EnhancedForLoopTree) targetStatementTree).getStatement();
+                    break;
+                case DO_WHILE_LOOP:
+                    if (((DoWhileLoopTree) targetStatementTree).getStatement().getKind() == Kind.BLOCK)
+                        block = ((DoWhileLoopTree) targetStatementTree).getStatement();
+                    break;
+            }
+        // No need to check for version of JDK, as targetExpressionTree can only be non-null in case of JDK-12 or higher
+        } else if (targetExpressionTree != null) {
+            try {
```

Comment:
Two comments:
-I guess I'd suggest to use TreeShims instead of directly using reflection. That would allow us to encapsulate the reflection on one place. It is not in java.editor.base yet, but simply pick an important class and a package in java.editor.base and add these into:
java/java.source.base/src/org/netbeans/modules/java/source/TreeShimsCopier.java

The TreeShims are not exactly pretty, but allow to keep the reflection stuff somewhat concentrated on one place. At least until/unless we find a better way to handle new JDK versions.

-but: what is the point here? Wouldn't body = targetExpressionTree work? After all, "body" is only used to find the closing '}'.